### PR TITLE
Add Composer Project Service (updated)

### PR DESCRIPTION
**Pull request by ikappas**
This commit adds the per-project option to expose a project as a composer package. It generates and maintains a packages.json file along with a project specific /p/project-{id}.json file (conforming to https://getcomposer.org/doc/05-repositories.md#includes) in the public/ folder, allowing the GitLab install to be used as a Composer repository.

```
Note: There is a known issue integrating with the GitLab-Omnibus, namely that the public folder is owned by 'root' and GitLab is run as 'git'. The GitLab-Omnibus chef script needs to be updated to allow access to both packages.json and the /p/ folder in the public/ folder
```

It extends on the original idea providing extended service configuration options that allow to work with packages that do not include a composer.json

It has three operational modes allowing for:
1. Default: Searches for a composer.json file in the source code root folder and if found exports the package.
2. Project: Parse project attributes to create a virtual composer.json to use for exporting the package.
3. Advanced: Parse a custom json string supplied in the service configuration to export the package.

It also has branch / tag filters that allows fine-grained filtering of specific branches/tags to be exposed as composer packages. 

In the process of providing a more complex configuration screen i added support for rendering fieldsets, and options in project service configuration along with enabling hints. Also notifications are enabled for Tag/Branch deletion from the gitlab host.

This is and updated merge request for https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/106 working with the latest source code.

Please see feature thread http://feedback.gitlab.com/forums/176466-general/suggestions/5953067-provide-composer-package-repository-support

When a project that has activated the composer service receives a push, the public /packages.json file and project specific json file will be refreshed to include version definitions for the branches and tags as specified in the service configuration in the project's repository.

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://ikappas-master.c68bfdc4cac4c32c41e1a922e73062ea.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://ikappas-master.c68bfdc4cac4c32c41e1a922e73062ea.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by rumpelsepp** [comment]
Sorry for this comment, but: What about running rubocop and fixing these errors locally? My github dashboard is full of hound messages caused by this pr.

**Comment by ikappas** [comment]
sorry about that, I just got around setting rubocop to fix these

**Comment by rumpelsepp** [comment]
Thx. :)
Usually you just have to run rubocop on the commandline and it shows the same errors. Houndci uses rubocop internally so it doesn't matter on which one you rely.

File: CHANGELOG [hunk 1]
```diff
@@ -16,64 +16,66 @@ v 7.8.0 (unreleased)
   - Improve sorting logic in UI and API. Explicitly define what sorting method is used by default
   - Allow more variations for commit messages closing issues (Julien Bianchi and Hannes Rosenögger)
   - Fix overflow at sidebar when have several itens
-  - 
+  -
```
Comment (Razer6) [inline CHANGELOG]: Can you please revert these changes?

**Comment by ikappas** [comment]
@Razer6 which changes specifically are you referring to?

File: CHANGELOG (hunk 1 above)
Comment (ikappas) [inline CHANGELOG]: @Razer6 which changes specifically are you referring to? the CHANGELOG?

**Comment by Razer6** [comment]
@ikappas The whitespace changes in the changelog? This PR is very huge. I propose factoring out the newly added DeleteTagService to a separate PR.

**Comment by ikappas** [comment]
Sure I can revert the changes to the changelog. 

As I am not that proficient on github I would  appreciate some help on how to go about factoring out the Delete Tag Service to a seperate PR?

**Comment by Razer6** [comment]
The `DeleteTagService` is done in 27b040d8a0aa26920360128724f8b55ae4fc53e0. I would go on the master branch, checkout to a new feature branch and cherry-pick this commit (`git cherry-pick 27b040d8a0aa26920360128724f8b55ae4fc53e0`) to the new branch. It would be awesome if you would add some tests to this service.

**Comment by achretien** [comment]
Is there any chance to see this feature in a future gitlab version?

**Comment by dosire** [comment]
@ikappas Can you make the build green maybe?

**Comment by ikappas** [comment]
Hi guys, I have been busy with other projects and just got some time to revisit this.
I have almost finished a partial port of composer php into a reusable ruby gem library so that gitlab can remain composer implementation agnostic. 
Also this pull request is quite outdated and quite large. Please close this pull request and I will make a new one with smaller changes.

**Comment by dosire** [comment]
Thanks @ikappas
